pkg/driver/controller.go

```diff
@@ -121,6 +121,13 @@ func (d *controllerService) CreateVolume(ctx context.Context, req *csi.CreateVol
 		return nil, status.Error(codes.InvalidArgument, errString)
 	}
 
+	// check if a request is already in-flight, move this block to the beginning of the request to protect idempotency of this API
+	if ok := d.inFlight.Insert(req.String()); !ok {
```

Comment:
what if we move this  all the way to beginning of the func, even before validating the request?

Also I  noticed it's inconsistent how some functions validate with volumeID but others with request...I think they should all use request?

Reply:
When I look into the spec, it is saying there is no more than one call “in-flight” **per volume** at a given time. I was struggled to determine using volumeId or whole req. If we use req, we are able to guarantee no more than one request per operation for one volume, but some case like deleteVolume & createSnapshot on same VolumeId will be allowed. Just wondering should we care about these use cases, otherwise use whole request make sense to me..

Reply:
let's consult the spec and think about it rigorously for each operation, I am scared of races :D

CreateVolume: We should use `name`: https://github.com/container-storage-interface/spec/blob/master/spec.md#controller-service-rpc
DeleteVolume: We should use `volume_id`: https://github.com/container-storage-interface/spec/blob/master/spec.md#deletevolume
CreateSnapshot: We should use `name`: https://github.com/container-storage-interface/spec/blob/master/spec.md#createsnapshot
DeleteSnapshot: We should use `snapshot_id`: https://github.com/container-storage-interface/spec/blob/master/spec.md#deletesnapshot

obviously, for createvolume and createsnapshot, you don't know the ID yet, so you can't use it as an idempotency token, that's why the spec requires `name`.